metaflow/plugins/cards/card_datastore.py

```diff
@@ -0,0 +1,251 @@
+        has_id = False
+        if len(file_split) == 4:
+            has_id = True
+        elif len(file_split) < 3 or len(file_split) > 4:
+            raise Exception(
+                "Invalid card file name %s. Card file names should be of form ID-TYPE-INDEX-HASH.html or TYPE-INDEX-HASH.html"
+                % card_file_name
+            )
+        card_id, card_type, card_index, card_hash = None, None, None, None
+        if has_id:
+            card_id, card_type, card_index, card_hash = file_split
+        else:
+            card_type, card_index, card_hash = file_split
+        card_hash = card_hash.split(".html")[0]
+        return card_id, card_type, card_index, card_hash
+
+    def save_card(self, card_type, card_id, card_index, card_html, overwrite=True):
+        card_file_name = self._make_card_file_name(
+            card_type, card_index, card_id=card_id
+        )
+        card_path = self.get_card_location(
+            self._get_card_path(), card_file_name, card_html
+        )
+        self._backend.save_bytes(
+            [(card_path, BytesIO(bytes(card_html, "utf-8")))], overwrite=overwrite
+        )
+
+    def _list_card_paths(
+        self, card_type=None, card_id=None, card_index=None, card_hash=None
+    ):
+        card_path = self._get_card_path()
+        card_paths = self._backend.list_content([card_path])
+        if len(card_paths) == 0:
+            # If there are no files found on the Path then raise an error of
+            raise CardNotPresentException(
+                self._flow_name,
+                self._run_id,
+                self._step_name,
+                card_index=card_index,
+                card_id=card_id,
+                card_hash=card_hash,
+                card_type=card_type,
+            )
+        cards_found = []
+        for task_card_path in card_paths:
+            card_path = task_card_path.path
+            cid, ctype, cidx, chash = self.card_info_from_path(card_path)
+            is_index_present = lambda x: True
+            if card_index is not None:
+                # if the index is not none then the lambda checks it
+                is_index_present = lambda x: x == cidx
+
+            if not is_index_present(str(card_index)):
+                continue
+
+            # if id,type or hash don't match then continue.
+            if card_id is not None and card_id != cid:
+                continue
+            elif card_type is not None and ctype != card_type:
+                continue
+            elif card_hash is not None:
+                if chash != card_hash and card_hash != chash[:NUM_SHORT_HASH_CHARS]:
+                    continue
+
+            if task_card_path.is_file:
+                cards_found.append(card_path)
+
+        return cards_found
+
+    def get_card_names(self, card_paths):
+        name_tuples = []
+        for path in card_paths:
+            name_tuples.append(self.card_info_from_path(path))
+        return name_tuples
```

Comment:
use list comprehension:
`  return [self.card_info_from_path(path) for path in card_paths]`